core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java

```diff
@@ -373,4 +434,196 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+    }
+
+    private static Map<String, Float> handleFieldsMatchPattern(MapperService mapperService, Map<String, Float> fieldsBoosts) {
+        Map<String, Float> newFieldsBoosts = new HashMap<>();
+        for (Map.Entry<String, Float> fieldBoost : fieldsBoosts.entrySet()) {
+            String fField = fieldBoost.getKey();
+            Float fBoost = fieldBoost.getValue();
+            if (Regex.isSimpleMatchPattern(fField)) {
+                for (String field : mapperService.simpleMatchToIndexNames(fField)) {
+                    newFieldsBoosts.put(field, fBoost);
+                }
+            } else {
+                newFieldsBoosts.put(fField, fBoost);
+            }
+        }
+        return newFieldsBoosts;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (value == null) {
+            validationException = addValidationError("no text specified for multi_match query.", validationException);
+        }
+        if (fieldsBoosts == null || fieldsBoosts.isEmpty()) {
+            validationException = addValidationError("no fields specified for multi_match query.", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected MultiMatchQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        MultiMatchQueryBuilder multiMatchQuery = new MultiMatchQueryBuilder(in.readGenericValue(),
+                (Map<String, Float>) in.readGenericValue());
+        if (in.readBoolean()) {
+            multiMatchQuery.type = Type.valueOf(in.readString());
+        }
+        if (in.readBoolean()) {
+            multiMatchQuery.operator = Operator.readOperatorFrom(in);
+        }
+        multiMatchQuery.analyzer = in.readOptionalString();
+        multiMatchQuery.slop = in.readOptionalVInt();
+        if (in.readBoolean()) {
+            multiMatchQuery.fuzziness = Fuzziness.readFuzzinessFrom(in);
+        }
+        multiMatchQuery.prefixLength = in.readOptionalVInt();
+        multiMatchQuery.maxExpansions = in.readOptionalVInt();
+        multiMatchQuery.minimumShouldMatch = in.readOptionalString();
+        multiMatchQuery.fuzzyRewrite = in.readOptionalString();
+        multiMatchQuery.useDisMax = in.readOptionalBoolean();
+        multiMatchQuery.tieBreaker = in.readOptionalFloat();
+        multiMatchQuery.lenient = in.readOptionalBoolean();
+        multiMatchQuery.cutoffFrequency = in.readOptionalFloat();
+        if (in.readBoolean()) {
+            multiMatchQuery.zeroTermsQuery = ZeroTermsQuery.valueOf(in.readString());
+        }
+        return multiMatchQuery;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeGenericValue(value);
+        out.writeGenericValue(fieldsBoosts);
+        if (type != null) {
+            out.writeBoolean(true);
+            out.writeString(type.name().toUpperCase(Locale.ROOT));
```

Comment:
Can we make MultiMatchQuery.Type a Writable similar to Operator etc...? Relying on serializing an ordinal feels safer to me than using the names.